sky/resources.py

```diff
@@ -558,6 +562,14 @@ def get_valid_regions_for_launchable(self) -> List[clouds.Region]:
 
         return filtered_regions
 
+    def set_docker_login_config(self, docker_login_config: Dict[str,
+                                                                str]) -> None:
+        if self.extract_docker_image() is None:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    'Docker login config is only supported for docker images.')
+        self._docker_login_config = docker_login_config
+
```

Comment:
`Resources` object is designed to be immutable. All the checks should be applied during the initialization, and modification should be done through `r.copy(_docker_login_config=xxx)`